src/Microsoft.AspNetCore.WebSockets/Internal/fx/src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs

```diff
@@ -183,7 +189,20 @@ private ManagedWebSocket(Stream stream, bool isServer, string subprotocol, TimeS
             _stream = stream;
             _isServer = isServer;
             _subprotocol = subprotocol;
-            _receiveBuffer = new byte[Math.Max(receiveBufferSize, MaxMessageHeaderLength)];
+
+            // If we were provided with a buffer to use, use it, as long as it's big enough for our needs, and for simplicity
+            // as long as we're not supposed to use only a portion of it.  If it doesn't meet our criteria, just create a new one.
+            if (receiveBuffer.HasValue &&
+                receiveBuffer.Value.Offset == 0 && receiveBuffer.Value.Count == receiveBuffer.Value.Array.Length &&
+                receiveBuffer.Value.Count >= MaxMessageHeaderLength)
+            {
+                _receiveBuffer = receiveBuffer.Value.Array;
+            }
+            else
+            {
+                _receiveBufferFromPool = true;
+                _receiveBuffer = ArrayPool<byte>.Shared.Rent(Math.Max(receiveBufferSize, MaxMessageHeaderLength));
```

Comment:
This change in particular, I'll be interested in knowing whether it helps/hurts scalability in your scenarios.  I've been tempted to revert it, out of a concern that we could be taking all of the buffers from the pool for long-lived websockets.